google/cloud/bigtable/internal/async_bulk_apply_test.cc

```diff
@@ -288,6 +285,179 @@ TEST_F(AsyncBulkApplyTest, AsyncBulkApplyFailureRetry) {
   EXPECT_TRUE(cq_impl_->empty());
 }
 
+TEST_F(AsyncBulkApplyTest, TooManyFailures) {
+  bigtable::BulkMutation mut{
+      bigtable::SingleRowMutation("foo2",
+                                  {bigtable::SetCell("f", "c", 0_ms, "v2")}),
+      bigtable::SingleRowMutation("foo3",
+                                  {bigtable::SetCell("f", "c", 0_ms, "v3")}),
+  };
+
+  // We give up on the 5th error.
+  const int limited_error_count = 4;
+
+  EXPECT_CALL(*client_, PrepareAsyncMutateRows)
+      .Times(limited_error_count + 1)
+      .WillRepeatedly([](grpc::ClientContext*,
+                         btproto::MutateRowsRequest const&,
+                         grpc::CompletionQueue*) {
+        auto reader = absl::make_unique<
+            MockClientAsyncReaderInterface<btproto::MutateRowsResponse>>();
+        EXPECT_CALL(*reader, Read).Times(2);
+        EXPECT_CALL(*reader, Finish).WillOnce([](grpc::Status* status, void*) {
+          *status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "try again");
+        });
+        EXPECT_CALL(*reader, StartCall);
+        return reader;
+      });
+
+  auto limited_retry_policy = LimitedErrorCountRetryPolicy(limited_error_count);
+  auto bulk_apply_future = internal::AsyncRetryBulkApply::Create(
+      cq_, limited_retry_policy.clone(), rpc_backoff_policy_->clone(),
+      *idempotent_mutation_policy_, metadata_update_policy_, client_,
+      "my-app-profile", "my-table", std::move(mut));
+
+  for (int retry = 0; retry < limited_error_count; ++retry) {
+    SimulateIterationThatRetries();
+    ASSERT_EQ(1U, cq_impl_->size());
+  }
+
+  SimulateIteration();
+
+  auto failures = bulk_apply_future.get();
+  EXPECT_EQ(2U, failures.size());
```

Comment:
Since these test were written we have gained a bunch of useful helpers, nowadays we would say:

```cc
EXPECT_THAT(failures, ElementsAre(StatusIs(StatusCode::kUnavailable), StatusIs(StatusCode::kUnavailable)));
```

On failure the expression I propose would be a lot more informative.

Reply:
Done.